src/TestSuite/IntegrationTestTrait.php

```diff
@@ -536,6 +536,11 @@ public function controllerSpy(EventInterface $event, ?Controller $controller = n
         }
         $this->_controller = $controller;
         $events = $controller->getEventManager();
+        $events->on('Controller.beforeRedirect', function () use ($controller): void {
+            if ($this->_retainFlashMessages) {
```

Comment:
```suggestion
        $events->on('Controller.beforeRedirect', function ($event): void {
            $controller = $event->getSubject();
            if ($this->_retainFlashMessages) {
```

Saves having another closure reference.